In [2]:
import newspaper
from newspaper import Article
import pandas as pd
import numpy as np


articles to ignore
https://www.nytimes.com/tips

In [16]:
def get_article(url):
    a = Article(url, language='en')
    a.download()
    a.parse()
    return a


In [4]:
ny = pd.read_csv('data/nytposts_readyforarticle.csv')

In [5]:
ny.shape

(2467, 10)

In [6]:
ny.head(1)

Unnamed: 0  index                      created_at             tweet_id  \
0           0      0  Mon Feb 18 05:49:55 +0000 2019  1097372619848986625   

                                                text                      url  \
0  Boys with high levels of inattention later ear...  https://nyti.ms/2EbrJZO   

   retweet_count                                             status  \
0             39  {"created_at": "Mon Feb 18 05:49:55 +0000 2019...   

                                      retweeter_list                final_url  
0  [486097436, 55651793, 191972027, 22681390, 897...  https://nyti.ms/2EbrJZO

In [ ]:
url = ny.final_url[0]
a = get_article(url)
print(a.text)
print(a.top_image)
print(a.title)
print(a.authors)
a.nlp()
print(a.summary)
print(a.keywords)

In [26]:

import time

file_path = 'data/art_info.csv'
df1 = pd.DataFrame(columns = ['tweet_id','text', 'top_image', 'title', 'authors', 'summary', 'keywords'])
df1.to_csv(file_path)

for i  in range(ny.shape[0]):
    t1 = time.time()
    print(i)
    url = ny.final_url[i]
    time.sleep(0.5)
    mini_df = pd.DataFrame()
    mini_df['tweet_id'] = [ny['tweet_id'][i]]
    
    try: 
        article = get_article(url)
        mini_df['text'] = [article.text]
        mini_df['top_image'] = [article.top_image]
        mini_df['title'] = [article.title]
        mini_df['authors'] = [article.authors]
        article.nlp()
        mini_df['summary'] = [article.summary]
        mini_df["keywords"] = [article.keywords]

    except Exception as e: 
        print('exception at index:', i, e)
        mini_df['text'] = np.nan
        mini_df['top_image'] = np.nan
        mini_df['title'] = np.nan
        mini_df['authors'] = np.nan
        article.nlp()
        mini_df['summary'] = np.nan
        mini_df["keywords"] = np.nan
    
    mini_df.to_csv(file_path, mode='a', header=False)
    t2 = time.time()
    print(t2-t1, 'seconds')

0
0.9418859481811523 seconds
1
1.2879889011383057 seconds
2
1.3580491542816162 seconds
3
1.154836893081665 seconds
4
1.440420150756836 seconds
5
1.4376919269561768 seconds
6
1.4471371173858643 seconds
7
1.2572681903839111 seconds
8
1.1856579780578613 seconds
9
1.1765260696411133 seconds
10
1.136962890625 seconds
11
1.1250498294830322 seconds
12
1.2353360652923584 seconds
13
1.5168969631195068 seconds
14
1.3704009056091309 seconds
15
1.3980772495269775 seconds
16
0.9895169734954834 seconds
17
1.3136372566223145 seconds
18
1.2061970233917236 seconds
19
1.049739122390747 seconds
20
1.0799241065979004 seconds
21
1.448481798171997 seconds
22
1.3960421085357666 seconds
23
1.5364458560943604 seconds
24
1.0156409740447998 seconds
25
1.1264350414276123 seconds
26
1.8219449520111084 seconds
27
1.1842658519744873 seconds
28
1.1256649494171143 seconds
29
1.1502912044525146 seconds
30
1.4360370635986328 seconds
31
1.1789979934692383 seconds
32
1.4242091178894043 seconds
33
1.4223737716674805 second

1.2645010948181152 seconds
272
1.4049429893493652 seconds
273
1.4576871395111084 seconds
274
1.4946918487548828 seconds
275
1.1586220264434814 seconds
276
1.519151210784912 seconds
277
2.3126060962677 seconds
278
1.467499017715454 seconds
279
1.431359052658081 seconds
280
1.6637170314788818 seconds
281
1.1711170673370361 seconds
282
1.2470228672027588 seconds
283
1.3571600914001465 seconds
284
1.4498958587646484 seconds
285
1.5474460124969482 seconds
286
1.5558090209960938 seconds
287
1.552847146987915 seconds
288
1.6033821105957031 seconds
289
1.4320271015167236 seconds
290
1.5092759132385254 seconds
291
1.2816948890686035 seconds
292
1.429589033126831 seconds
293
1.3914637565612793 seconds
294
1.1625280380249023 seconds
295
1.073503017425537 seconds
296
1.1365208625793457 seconds
297
1.0244011878967285 seconds
298
1.4986159801483154 seconds
299
1.4629130363464355 seconds
300
1.4855120182037354 seconds
301
1.1362190246582031 seconds
302
1.2441670894622803 seconds
303
1.127928018569946

1.494483232498169 seconds
539
1.396521806716919 seconds
540
1.4224140644073486 seconds
541
1.2650630474090576 seconds
542
1.454543113708496 seconds
543
1.201219081878662 seconds
544
1.253427267074585 seconds
545
1.1924889087677002 seconds
546
0.9474267959594727 seconds
547
1.4678966999053955 seconds
548
1.1073529720306396 seconds
549
1.0073001384735107 seconds
550
0.9905328750610352 seconds
551
1.4900078773498535 seconds
552
1.176875114440918 seconds
553
1.3995780944824219 seconds
554
1.3956489562988281 seconds
555
1.2710871696472168 seconds
556
1.1247899532318115 seconds
557
1.412994146347046 seconds
558
1.1060810089111328 seconds
559
1.0908379554748535 seconds
560
1.9186170101165771 seconds
561
1.0853800773620605 seconds
562
1.4203650951385498 seconds
563
1.2308681011199951 seconds
564
1.1070148944854736 seconds
565
1.167809009552002 seconds
566
1.115443229675293 seconds
567
1.4154491424560547 seconds
568
1.3818988800048828 seconds
569
1.3857090473175049 seconds
570
1.222563743591308

1.8567068576812744 seconds
807
1.4217190742492676 seconds
808
0.9426999092102051 seconds
809
1.4570960998535156 seconds
810
1.4589009284973145 seconds
811
1.4312069416046143 seconds
812
1.2743501663208008 seconds
813
1.1573669910430908 seconds
814
1.142406940460205 seconds
815
1.0957069396972656 seconds
816
1.3599600791931152 seconds
817
1.3807499408721924 seconds
818
1.4700102806091309 seconds
819
1.2459559440612793 seconds
820
1.1412019729614258 seconds
821
1.4172728061676025 seconds
822
1.2145028114318848 seconds
823
1.1129271984100342 seconds
824
1.4433059692382812 seconds
825
1.412405014038086 seconds
826
1.3933560848236084 seconds
827
1.3600380420684814 seconds
828
1.2909140586853027 seconds
829
0.968134880065918 seconds
830
1.1667866706848145 seconds
831
0.9308810234069824 seconds
832
1.4773259162902832 seconds
833
1.4159607887268066 seconds
834
1.3457529544830322 seconds
835
1.1361887454986572 seconds
836
1.4427220821380615 seconds
837
1.1214449405670166 seconds
838
0.968697071

1.1944348812103271 seconds
1071
1.435711145401001 seconds
1072
1.2031879425048828 seconds
1073
1.2839796543121338 seconds
1074
0.9561328887939453 seconds
1075
1.442798137664795 seconds
1076
1.2442529201507568 seconds
1077
2.8169870376586914 seconds
1078
2.126840829849243 seconds
1079
1.7936630249023438 seconds
1080
1.3789477348327637 seconds
1081
1.418313980102539 seconds
1082
1.960062026977539 seconds
1083
1.1345829963684082 seconds
1084
1.1669597625732422 seconds
1085
1.157806158065796 seconds
1086
1.23826003074646 seconds
1087
1.4010732173919678 seconds
1088
1.8377060890197754 seconds
1089
1.3888881206512451 seconds
1090
1.3910701274871826 seconds
1091
1.7221040725708008 seconds
1092
1.7302188873291016 seconds
1093
1.2203168869018555 seconds
1094
1.4297189712524414 seconds
1095
1.25636625289917 seconds
1096
1.4259169101715088 seconds
1097
1.28249192237854 seconds
1098
1.4360589981079102 seconds
1099
1.0946860313415527 seconds
1100
0.9565727710723877 seconds
1101
1.5973527431488037 s

1.1101429462432861 seconds
1330
1.3581218719482422 seconds
1331
2.2944841384887695 seconds
1332
1.8690228462219238 seconds
1333
1.361677885055542 seconds
1334
1.5675339698791504 seconds
1335
2.5830001831054688 seconds
1336
1.2450199127197266 seconds
1337
1.514822006225586 seconds
1338
1.4694480895996094 seconds
1339
1.2469680309295654 seconds
1340
1.1937427520751953 seconds
1341
1.8147659301757812 seconds
1342
1.5546848773956299 seconds
1343
1.5090110301971436 seconds
1344
2.1236231327056885 seconds
1345
1.45796799659729 seconds
1346
1.163362979888916 seconds
1347
1.1140501499176025 seconds
1348
0.9607930183410645 seconds
1349
1.5028719902038574 seconds
1350
2.3013429641723633 seconds
1351
1.9293460845947266 seconds
1352
3.0986979007720947 seconds
1353
1.3980722427368164 seconds
1354
1.3277850151062012 seconds
1355
1.6552391052246094 seconds
1356
1.4013311862945557 seconds
1357
1.2484538555145264 seconds
1358
1.345940351486206 seconds
1359
1.5035829544067383 seconds
1360
1.415083885192

1.2566189765930176 seconds
1589
1.8706529140472412 seconds
1590
1.1852118968963623 seconds
1591
2.0847463607788086 seconds
1592
1.3813073635101318 seconds
1593
1.3964228630065918 seconds
1594
1.3985567092895508 seconds
1595
1.3958070278167725 seconds
1596
1.2827179431915283 seconds
1597
1.1365089416503906 seconds
1598
1.9711320400238037 seconds
1599
1.8572559356689453 seconds
1600
1.9460086822509766 seconds
1601
1.4715430736541748 seconds
1602
1.5223419666290283 seconds
1603
2.1408817768096924 seconds
1604
1.4163641929626465 seconds
1605
2.196046829223633 seconds
1606
1.1419572830200195 seconds
1607
1.1590518951416016 seconds
1608
1.1872990131378174 seconds
1609
1.5220301151275635 seconds
1610
2.4836909770965576 seconds
1611
1.448498249053955 seconds
1612
1.1951289176940918 seconds
1613
1.8458998203277588 seconds
1614
1.7394239902496338 seconds
1615
1.40580415725708 seconds
1616
1.164708137512207 seconds
1617
1.4549047946929932 seconds
1618
1.4110839366912842 seconds
1619
1.46615600585

1.9238719940185547 seconds
1848
1.8797709941864014 seconds
1849
1.9816019535064697 seconds
1850
1.413226842880249 seconds
1851
1.4484388828277588 seconds
1852
1.589388132095337 seconds
1853
1.5103609561920166 seconds
1854
2.116194725036621 seconds
1855
2.124800682067871 seconds
1856
1.7875587940216064 seconds
1857
1.1889569759368896 seconds
1858
1.214141845703125 seconds
1859
1.2785470485687256 seconds
1860
1.4063639640808105 seconds
1861
1.7051401138305664 seconds
1862
2.7303168773651123 seconds
1863
2.1088991165161133 seconds
1864
2.118577003479004 seconds
1865
2.1250720024108887 seconds
1866
1.38018798828125 seconds
1867
1.938701868057251 seconds
1868
1.121553897857666 seconds
1869
1.5966320037841797 seconds
1870
0.9560401439666748 seconds
1871
1.3644449710845947 seconds
1872
1.1301302909851074 seconds
1873
2.4344067573547363 seconds
1874
1.1298060417175293 seconds
1875
1.6333496570587158 seconds
1876
1.5655791759490967 seconds
1877
1.4243659973144531 seconds
1878
3.2667369842529297

1.1878361701965332 seconds
2107
1.1503660678863525 seconds
2108
2.029541015625 seconds
2109
1.1613829135894775 seconds
2110
1.1213228702545166 seconds
2111
2.0565152168273926 seconds
2112
1.355163812637329 seconds
2113
1.3894407749176025 seconds
2114
1.3524620532989502 seconds
2115
1.3165082931518555 seconds
2116
1.4457950592041016 seconds
2117
2.0202078819274902 seconds
2118
1.7704789638519287 seconds
2119
1.0483691692352295 seconds
2120
1.224062204360962 seconds
2121
1.534911870956421 seconds
2122
1.3631830215454102 seconds
2123
1.360590934753418 seconds
2124
2.3191652297973633 seconds
2125
2.5915379524230957 seconds
2126
2.34426212310791 seconds
2127
1.089555025100708 seconds
2128
0.9656667709350586 seconds
2129
1.0899240970611572 seconds
2130
1.1033899784088135 seconds
2131
1.2038028240203857 seconds
2132
1.4293220043182373 seconds
2133
1.429548978805542 seconds
2134
2.171880006790161 seconds
2135
1.1456048488616943 seconds
2136
1.50465989112854 seconds
2137
1.447248935699463 secon

1.8800148963928223 seconds
2366
1.4712729454040527 seconds
2367
1.3971490859985352 seconds
2368
1.8759939670562744 seconds
2369
1.184345006942749 seconds
2370
2.511212110519409 seconds
2371
1.556274175643921 seconds
2372
1.4539058208465576 seconds
2373
1.0736041069030762 seconds
2374
1.1996629238128662 seconds
2375
1.1378731727600098 seconds
2376
1.6340076923370361 seconds
2377
1.1708190441131592 seconds
2378
1.4913389682769775 seconds
2379
1.3849961757659912 seconds
2380
1.2037460803985596 seconds
2381
1.3557980060577393 seconds
2382
1.8790709972381592 seconds
2383
1.4510858058929443 seconds
2384
1.4841129779815674 seconds
2385
2.1203057765960693 seconds
2386
1.136139154434204 seconds
2387
1.736983060836792 seconds
2388
2.2039530277252197 seconds
2389
1.467332124710083 seconds
2390
1.2746849060058594 seconds
2391
1.1630971431732178 seconds
2392
1.199328899383545 seconds
2393
1.548201084136963 seconds
2394
1.8532443046569824 seconds
2395
1.4123709201812744 seconds
2396
1.01134514808654

In [28]:
art_info = pd.read_csv(file_path)

In [39]:
nan_texts = art_info[art_info.text.isna()]
print(nan_texts.shape)
nan_texts.head(2)

(33, 8)


Unnamed: 0             tweet_id text  \
24           0  1097254401847246849  NaN   
38           0  1097205320395489281  NaN   

                                            top_image  \
24  https://static01.nyt.com/newsgraphics/2018/11/...   
38  https://static01.nyt.com/newsgraphics/2018/11/...   

                                   title authors summary  \
24  Where to go in 2019: Dessau, Germany      []     NaN   
38  Where to go in 2019: Calgary, Canada      []     NaN   

                         keywords  
24  ['germany', '2019', 'dessau']  
38  ['calgary', 'canada', '2019']

In [40]:
art_info.drop(nan_texts.index, inplace = True)
art_info.reset_index(inplace = True)
art_info.head(1)

index  Unnamed: 0             tweet_id  \
0      0           0  1097372619848986625   

                                                text  \
0  A kindergarten boy’s behavior could predict hi...   

                                           top_image  \
0  https://static01.nyt.com/images/2016/09/14/wel...   

                                        title               authors  \
0  Behavior at Age 6 May Predict Adult Income  ['Nicholas Bakalar']   

                                             summary  \
0  A kindergarten boy’s behavior could predict hi...   

                                            keywords  
0  ['boys', 'income', 'inattention', 'age', 'beha...

In [46]:
# art_info.drop(['index', 'Unnamed: 0'], axis =1, inplace = True)
print(art_info.shape)
art_info.head(2)

(2434, 7)


tweet_id                                               text  \
0  1097372619848986625  A kindergarten boy’s behavior could predict hi...   
1  1097368941842837506  In more than 14 years of roaming Mars, NASA’s ...   

                                           top_image  \
0  https://static01.nyt.com/images/2016/09/14/wel...   
1  https://static01.nyt.com/images/2019/02/15/sci...   

                                               title               authors  \
0         Behavior at Age 6 May Predict Adult Income  ['Nicholas Bakalar']   
1  Photos From the Opportunity Rover’s Mission on...     ['Kenneth Chang']   

                                             summary  \
0  A kindergarten boy’s behavior could predict hi...   
1  In more than 14 years of roaming Mars, NASA’s ...   

                                            keywords  
0  ['boys', 'income', 'inattention', 'age', 'beha...  
1  ['mission', 'rovers', 'mars', 'opportunity', '...

In [47]:
art_info.to_csv('data/art_info1.csv')

# Summary: 

1. No of articles 2434
2. Next head to all_users_info.ipynb to make list of all users and their details 
3. Final output file 'data/art_info1.csv)
